pandas/core/reshape/merge.py

```diff
@@ -809,10 +809,18 @@ def _create_join_index(self, index, other_index, indexer,
                 # if values missing (-1) from target index,
                 # take from other_index instead
                 join_list = join_index.to_numpy()
-                other_list = other_index.take(other_indexer).to_numpy()
-                join_list[mask] = other_list[mask]
-                join_index = Index(join_list, dtype=join_index.dtype,
-                                   name=join_index.name)
+                try:
+                    other_list = other_index.take(other_indexer).to_numpy()
+                    join_list[mask] = other_list[mask]
+                    join_index = Index(join_list, dtype=other_index.dtype,
+                                       name=other_index.name)
+                except ValueError:
```

Comment:
we *really* don't want to do a try/except here. What is falling into the except?

Reply:
can you:

- always just do this (what you have in the except)
- or use ``is_dtype_equal`` to test?

Reply:
I don't think `is_dtype_equal` would work because some combinations of different dtypes are still usable.

I agree with the first option because joining on the index of the other frame kind of makes the row order of the other frame arbitrary anyway.